third_party/hexagon/svdf_int8.cc

```diff
@@ -0,0 +1,106 @@
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER
+OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+==============================================================================*/
+
+#include <math.h>
+
+#include "hexagon_tflm_translation_svdf.h"
+#include "tensorflow/lite/c/builtin_op_data.h"
+#include "tensorflow/lite/c/common.h"
+#include "tensorflow/lite/kernels/internal/common.h"
+#include "tensorflow/lite/kernels/internal/quantization_util.h"
+#include "tensorflow/lite/kernels/internal/tensor_ctypes.h"
+#include "tensorflow/lite/kernels/kernel_util.h"
+#include "tensorflow/lite/kernels/op_macros.h"
+#include "tensorflow/lite/micro/kernels/activation_utils.h"
+#include "tensorflow/lite/micro/kernels/hexagon/hexagon_svdf.h"
+#include "tensorflow/lite/micro/kernels/kernel_util.h"
+#include "tensorflow/lite/micro/micro_utils.h"
+
+namespace tflite {
+
+TfLiteStatus SvdfEvalInt8(TfLiteContext* context, TfLiteNode* node) {
+  auto* params = reinterpret_cast<TfLiteSVDFParams*>(node->builtin_data);
+  TFLITE_DCHECK(node->user_data != nullptr);
+  const HexagonOpDataSvdf& data =
+      *(static_cast<const HexagonOpDataSvdf*>(node->user_data));
+
+  const TfLiteEvalTensor* input =
+      tflite::micro::GetEvalInput(context, node, kSvdfInputTensor);
+  const TfLiteEvalTensor* weights_feature =
+      tflite::micro::GetEvalInput(context, node, kSvdfWeightsFeatureTensor);
+  const TfLiteEvalTensor* weights_time =
+      tflite::micro::GetEvalInput(context, node, kSvdfWeightsTimeTensor);
+  const TfLiteEvalTensor* bias =
+      (NumInputs(node) == 5)
+          ? tflite::micro::GetEvalInput(context, node, kSvdfBiasTensor)
+          : nullptr;
+  TfLiteEvalTensor* activation_state = tflite::micro::GetMutableEvalInput(
+      context, node, kSvdfInputActivationStateTensor);
+  TfLiteEvalTensor* output =
+      tflite::micro::GetEvalOutput(context, node, kSvdfOutputTensor);
+
+  if (tflite::hexagon_svdf::HexagonOptimizable(context, node)) {
+    tflite::hexagon_svdf::HexagonEvalIntegerSVDF(
+        context, node, input, weights_feature, weights_time, bias, params,
+        activation_state, output, node->user_data);
+  } else {
+    EvalIntegerSvdfReference(context, node, input, weights_feature,
+                             weights_time, bias, params, activation_state,
+                             output, data.reference_op_data);
+  }
+  return kTfLiteOk;
+}
+
+TfLiteRegistration Register_SVDF_INT8() {
+  return {/*init=*/SvdfInit,
+          /*free=*/nullptr,
+          /*prepare=*/SvdfPrepare,
```

Comment:
***IF*** the Hexagon toolchain has the same linker issues as the Xtensa toolchain (that symbols don't get dropped from within a .cc file), then we should not have any dependency from svdf_int8.cc to svdf.cc

What that would mean is that the Init and Prepare functions should be moved to this file and svdf.cc should use the Init defined in svdf_int8.cc

Since we are being extra careful and separating the variants into their own .cc files, let's also move the init and prepare functions.